pytorch_lightning/utilities/apply_func.py

```diff
@@ -117,10 +117,10 @@ def apply_to_collection(
         return elem_type(*out) if is_namedtuple else elem_type(out)
 
     if _is_dataclass_instance(data):
-        out_dict = {}
-        for field in data.__dataclass_fields__:
+        result = copy(data)
```

Comment:
Could the copy be avoided?

Reply:
Probably, although it might reduce the readability of the code by using lower-level operations. I think `elem_type.__new__()` could be used to instantiate an instance of the same dataclass, but I'd have to test that. 

This is at least a shallow copy, so it is pretty lightweight. Let me know what your preference is.

Reply:
Cool. I didn't realize a `deepcopy` was acceptable. Not all tests run locally for me, but I think the latest version resolves the problem.